File: bindings/rust/examples/logging_demo.rs
use sop::{Context, Database, DatabaseOptions, LogLevel, manage_logging};
use std::fs;
use std::path::Path;

fn main() {
    println!("--- Logging Demo ---");

    // 1. Configure Logging
    let log_file = "sop_demo.log";
    if Path::new(log_file).exists() {
        fs::remove_file(log_file).unwrap();
    }

    println!("Configuring logger to write to {}...", log_file);
    if let Err(e) = manage_logging(LogLevel::Debug, log_file) {
        eprintln!("Failed to configure logging: {}", e);
        return;
    }

    // 2. Initialize Context & Database
    let ctx = Context::new();
    let db_path = "data/logging_demo_db";
    if Path::new(db_path).exists() {
        fs::remove_dir_all(db_path).unwrap();
    }

    println!("Opening database at {}...", db_path);
    let db = Database::new(&ctx, DatabaseOptions {
        stores_folders: Some(vec![db_path.to_string()]),
        ..Default::default()
    }).unwrap();

    // 3. Perform Operations
    println!("Starting transaction...");
    let trans = db.begin_transaction(&ctx).unwrap();
    
    println!("Creating B-Tree...");
    let btree = db.new_btree::<String, String>(&ctx, "logging_btree", &trans, None).unwrap();

    println!("Adding item...");
    btree.add(&ctx, "hello".to_string(), "world".to_string()).unwrap();

    println!("Committing transaction...");
    trans.commit(&ctx).unwrap();

    // 4. Verify Logs
    if Path::new(log_file).exists() {
        println!("\nSuccess! Log file created at {}.", log_file);
        println!("First 5 lines of log:");
        let content = fs::read_to_string(log_file).unwrap();
        for (i, line) in content.lines().enumerate() {
            if i >= 5 { break; }
            println!("{}", line);
        }
    } else {
        println!("Error: Log file was not created.");
    }
    
    println!("--- End of Logging Demo ---");
}

File: bindings/python/sop/examples/logging_demo.py
import os
import sys

# Add the parent directory to sys.path to import sop
# sys.path.append(os.path.join(os.path.dirname(__file__), '..'))

from sop import Context, Logger, LogLevel
from sop.ai import Database, DatabaseType, Item as VectorItem
from sop.database import DatabaseOptions

def main():
    # 1. Configure Logging
    # We'll log to a file named 'sop_app.log' with DEBUG level to see all details.
    log_file = "sop_app.log"
    print(f"Configuring logger to write to {log_file}...")
    Logger.configure(LogLevel.Debug, log_file)

    # 2. Initialize Context & Database
    ctx = Context()
    db_path = "data/logging_demo_db"
    
    print(f"Opening database at {db_path}...")
    # This operation will generate logs on the Go side
    db = Database(DatabaseOptions(stores_folders=[db_path], type=DatabaseType.Standalone))

    # 3. Perform Operations
    print("Starting transaction...")
    with db.begin_transaction(ctx) as tx:
        
        # Open a Vector Store
        print("Opening Vector Store...")
        vs = db.open_vector_store(ctx, tx, "demo_vectors")
        
        # Upsert an item
        print("Upserting item...")
        vs.upsert(ctx, VectorItem(
            id="item_1",
            vector=[0.5, 0.5, 0.5],
            payload={"description": "A test item"}
        ))
        
        # The commit happens automatically at the end of the 'with' block
        print("Committing transaction...")

    # 4. Verify Logs
    if os.path.exists(log_file):
        print(f"\nSuccess! Log file created at {log_file}.")
        print("First 5 lines of log:")
        with open(log_file, 'r') as f:
            for i, line in enumerate(f):
                if i >= 5: break
                print(line.strip())
    else:
        print("Error: Log file was not created.")

if __name__ == "__main__":
    main()
